### PR TITLE
:recycle: refactor(FileStreamInputArchiver): Utilize FileUtils for context-aware path usage.

### PR DESCRIPTION
**Pull request by Okanmercan99**

### CODE REVIEW

File: tofhir-engine/src/main/scala/io/tofhir/engine/execution/FileStreamInputArchiver.scala [hunk 1]
```diff
@@ -48,19 +48,22 @@ class FileStreamInputArchiver(runningJobRegistry: RunningJobRegistry) {
   def applyArchivingOnStreamingJob(taskExecution: FhirMappingJobExecution, mappingUrl: String): Unit = {
     // Get the commit file directory for this execution
     val checkPointDirectory: String = taskExecution.getCheckpointDirectory(mappingUrl)
-    val commitFileDirectory: File = Paths.get(checkPointDirectory, "commits").toFile
+    val commitFile: File = FileUtils.getPath(checkPointDirectory, "commits").toFile
```
Comment (suatgonul) [inline tofhir-engine/src/main/scala/io/tofhir/engine/execution/FileStreamInputArchiver.scala:51]: This is supposed to be a directory why did you change it to "commitfile"?

File: tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    when(mockTaskExecution.id).thenReturn(jobId)
+    when(mockTaskExecution.job).thenReturn(mockJob)
+    when(mockJob.dataProcessingSettings).thenReturn(mockDataProcessingSettings)
+    when(mockDataProcessingSettings.archiveMode).thenReturn(ArchiveModes.ARCHIVE)
+    when(mockTaskExecution.getCheckpointDirectory(mappingUrl)).thenReturn(
+      FileUtils.getPath("test-archiver", jobId, mappingUrl.hashCode.toString).toString)
+
+
+    // Create a source file to refer location of test.csv
+    val sourceFile = FileUtils.getPath("test-archiver", jobId, mappingUrl.hashCode.toString, "sources", "0", "0").toFile
```
Comment (suatgonul) [inline tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala:35]: We can extract the logic to access to source folder of Spark and add a dedicated test for that

File: tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    // Create a commit file inorder to start range function in applyArchivingOnStreamingJob
+    val commitFile = FileUtils.getPath("test-archiver", jobId, mappingUrl.hashCode.toString, "commits", "0").toFile
+    // Ensure the parent directories exist, if not, create them
+    commitFile.getParentFile.mkdirs()
+    val commitWriter = new PrintWriter(commitFile)
+    commitWriter.write("v1\n{\"nextBatchWatermarkMs\":0}")
+    commitWriter.close()
+
+    // Find the relative path between the workspace folder and the file to be archived
+    val relPath = FileUtils.getPath("").toAbsolutePath.relativize(FileUtils.getPath("test-archiver", "test.csv").toAbsolutePath)
```
Comment (suatgonul) [inline tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala:60]: Can you validate the value of "relPath"?

File: tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    // Call archiving function
+    fileStreamInputArchiver.applyArchivingOnStreamingJob(mockTaskExecution, mappingUrl)
+
+    // Check whether archiving file exists and csv file is moved
+    finalArchivePath.toFile.exists() shouldBe true
+    testCsvFile.exists() shouldBe false
+
+    // Clean test directory
+    org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.getPath("test-archiver").toFile)
+    org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.getPath(ToFhirConfig.engineConfig.archiveFolder).toFile)
```
Comment (suatgonul) [inline tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala:77]: archiveFolder takes the context path into account already. Why did you call FileUtils.getPath again?

File: tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+import io.tofhir.engine.execution.{FileStreamInputArchiver, RunningJobRegistry}
+import io.tofhir.engine.model._
+import io.tofhir.engine.util.FileUtils
+import org.mockito.MockitoSugar._
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+import java.io.PrintWriter
+
+class FileStreamInputArchiverTest extends AnyFlatSpec with Matchers {
```
Comment (suatgonul) [inline tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala:13]: Can you add tests for :
- delete mode
- multiple files
- validating the logic of getLastCommitOffset
- validating that only the recently committed files are archived but not the ones that are already archived

**Review by suatgonul — CHANGES_REQUESTED** [review]

File: tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala (hunk 4 above)
Comment (Okanmercan99) [reply tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala:77]: because archiveFolder is not a path, it is a string.

File: tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+  val runningJobRegistryMock = mock[RunningJobRegistry]
+  val fileStreamInputArchiver = new FileStreamInputArchiver(runningJobRegistryMock)
+
+  "FileStreamInputArchiver" should "apply archiving for a streaming job" in {
+
+    // Create a mock execution
+    val mockTaskExecution = mock[FhirMappingJobExecution]
```
Comment (suatgonul) [inline tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala:23]: Can you convert this to real object instead of mocking, as you are duplicating the logic with directories, which is error-prone and makes maintenance harder.

File: tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+    val testExecution = FhirMappingJobExecution(job = mockJob)
+
+    // Test whether commit directory is right
+    testExecution.getCommitDirectory(mappingUrl) shouldBe
+      FileUtils.getPath(ToFhirConfig.sparkCheckpointDirectory, jobId, mappingUrl.hashCode.toString, "commits").toString
+  }
+
+  "FileStreamInputArchiver" should "get source file" in {
+    // Create a mock job
+    val mockJob = mock[FhirMappingJob]
```
Comment (suatgonul) [inline tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala:199]: Let's apply mocking when there is no way to instantiate real objects. In other words, we should use real objects, if possible e.g. in this case

File: tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+    FileStreamInputArchiver.applyArchivingOnBatchJob(mockTaskExecution)
+
+    // Check whether input file is deleted
+    inputFile.exists() shouldBe false
+
+    // Clean test directories
+    org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.getPath(ToFhirConfig.engineConfig.contextPath, sourceFolderPath).toFile)
+  }
+
+  "FileStreamInputArchiver" should "get commit file" in {
```
Comment (suatgonul) [inline tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala:182]: This test and the following one are not belong to this test class. Can you create a dedicated test class?

File: tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+      FileUtils.getPath(ToFhirConfig.sparkCheckpointDirectory, jobId, mappingUrl.hashCode.toString, "sources", "0").toString
+  }
+
+  "FileStreamInputArchiver" should "get input files" in {
+
+    val mappingUrl = "mocked_mapping_url"
+    val jobId = "mocked_job_id_3"
+
+    // Create a source file to refer location of test.csv
+    val sourceFile = FileUtils.getPath("test-archiver", jobId, mappingUrl.hashCode.toString, "sources", "0", "0").toFile
```
Comment (suatgonul) [inline tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala:218]: Creating path with "sources", "0" is duplicated in many places. This logic should be implemented in a single place such as "SparkUtil". This would ease for instance handling a change in Spark's way of running.

File: tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+    // Path of test.csv and test2.csv
+    val testCsvFile = FileUtils.getPath("test-archiver", "test.csv").toFile
+    val test2CsvFile = FileUtils.getPath("test-archiver", "test2.csv").toFile
+    // Ensure the parent directories exist, if not, create them
+    sourceFile.getParentFile.mkdirs()
+    testCsvFile.getParentFile.mkdirs()
+    test2CsvFile.getParentFile.mkdirs()
+    // Write path of test.csv and test2.csv to the source file
+    val sourceWriter = new PrintWriter(sourceFile)
+    sourceWriter.write(s"{\"path\":\"${testCsvFile.getAbsolutePath.replace("\\", "\\\\")}\"}\n")
```
Comment (suatgonul) [inline tofhir-engine/src/test/scala/io/tofhir/test/engine/execution/FileStreamInputArchiverTest.scala:228]: Again, such Spark-specific logic should be maintained in a single place

**Review by suatgonul — CHANGES_REQUESTED** [review]

**Review by suatgonul — APPROVED** [review]
